clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -208,43 +290,88 @@ public RecordAppendResult append(TopicPartition tp,
                 return new RecordAppendResult(null, false, false, true);
             }
 
-            byte maxUsableMagic = apiVersions.maxUsableProduceMagic();
-            int size = Math.max(this.batchSize, AbstractRecords.estimateSizeInBytesUpperBound(maxUsableMagic, compression, key, value, headers));
-            log.trace("Allocating a new {} byte message buffer for topic {} partition {} with remaining timeout {}ms", size, tp.topic(), tp.partition(), maxTimeToBlock);
-            buffer = free.allocate(size, maxTimeToBlock);
+            RecordAppendResult appendResult = appendNewBatch(topic, partition, dq, sizeStats, timestamp, key, value, headers, callbacks, maxTimeToBlock);
+
+            // To properly estimate the amount of bytes produced to a partition, we keep track
+            // of batch headers.  The record size estimator would atomically extract this value
+            // and account for it in the record byte estimation.
+            if (appendResult.newBatchCreated && countBatchHeader)
+                sizeStats.onNewBatch(apiVersions.maxUsableProduceMagic(), compression);
 
+            return appendResult;
+        } finally {
+            appendsInProgress.decrementAndGet();
+        }
+    }
+
+    /**
+     * Append a new batch to the queue
+     *
+     * @param topic The topic
+     * @param partition The partition (cannot be RecordMetadata.UNKNOWN_PARTITION)
+     * @param dq The queue
+     * @param sizeStats The size stats
+     * @param timestamp The timestamp of the record
+     * @param key The key for the record
+     * @param value The value for the record
+     * @param headers the Headers for the record
+     * @param callbacks The callbacks to execute
+     * @param maxTimeToBlock The maximum time in milliseconds to block for buffer memory to be available
+     */
+    private RecordAppendResult appendNewBatch(String topic,
```

Comment:
This is mostly just moving part of the logic from append to a separate function.